### PR TITLE
C++ generation string signal

### PR DESCRIPTION
**Pull request by alisonbelow**
Partial template specialization for Frame class when PhysicalDataType is std::string. This allows us to get/set std::string values for Signals.

Would like feedback on conversion methods for `uint*` <-> `std::string` - more efficient or better approach?

### CODE REVIEW

File: cantools/database/can/cpp_source.py [hunk 1]
```diff
@@ -226,6 +225,13 @@ def _format_comment_no_tabs(comment):
         return ''
 
 
+# TODO this should evaluate outputs to integers/bools as well, just not only std::string or double
+def _signal_physical_type(signal):
```
Comment (avinashmittur) [inline cantools/database/can/cpp_source.py:229]: Dumb question that you've probably already thought about, but does the data structure that `signal` belongs to not have a field that we can run `type()`  function on?

Comment (avinashmittur) [reply cantools/database/can/cpp_source.py:229]: Disregard, didn't think this through.

**Review by avinashmittur — APPROVED** [review]

File: cantools/database/can/cpp_source.py [hunk 2]
```diff
@@ -226,6 +225,13 @@ def _format_comment_no_tabs(comment):
         return ''
 
 
+# TODO this should evaluate outputs to integers/bools as well, just not only std::string or double
```
Comment (alschult-saildrone) [inline cantools/database/can/cpp_source.py:228]: What is the reason not to just do that now, is it harder for some reason?

File: cantools/database/can/templates/DBC.h [hunk 3]
```diff
@@ -234,4 +236,78 @@ class Signal {
   uint32_t spn_ {J1939_INVALID_SPN};
 };
 
+/** Partial template specialization for std::string PhysicalDataType */
+template<typename RawDataType>
+class Signal<RawDataType, std::string> {
+ public:
+  Signal(const uint8_t* buffer, const std::string& name)
```
Comment (alschult-saildrone) [inline cantools/database/can/templates/DBC.h:243]: This gives you the buffer but not the length?

File: cantools/database/can/templates/DBC.h [hunk 4]
```diff
@@ -234,4 +236,78 @@ class Signal {
...
+  bool InRange(const std::string& value) const {
+    RawDataType raw = Encode(value);
+    return RawInRange(raw);
+  }
+
+  /** Decode given signal by applying scaling and offset. */
+  std::string Decode(RawDataType value) const {
+    unsigned char char_arr[sizeof(value)];
+    std::memcpy(char_arr, &value, sizeof(value));
+    std::string out(reinterpret_cast<const char*>(char_arr), sizeof(char_arr) / sizeof(char_arr[0]));
```
Comment (alschult-saildrone) [inline cantools/database/can/templates/DBC.h]: I'm not sure why you need the memcpy to the `char_arr`, could you not just do the same reinterpret cast when you instantiate the string from the incoming `value`?

File: tests/files/dbc/string_signals.dbc [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	SG_MUL_VAL_
+
+BS_:
+
+BU_:
+
+
+BO_ 10 PersonName: 40 Vector___XXX
+ SG_ first_name : 0|160@1+ (1,0) [0|0] "String" Vector__XXX
+ SG_ last_name : 160|160@1+ (1,0) [0|0] "str" Vector__XXX
```
Comment (alschult-saildrone) [inline tests/files/dbc/string_signals.dbc]: You might want to have a test case that includes some other types in it as well to make sure there is no offset issues.  I'd suggest a test that has all types (or if you already have that, add strings to that test case).

**Review by alschult-saildrone — APPROVED** [review]

File: cantools/database/can/cpp_source.py (hunk 2 above)
Comment (alisonbelow) [reply cantools/database/can/cpp_source.py:228]: We need to add `SIG_VALTYPE_` attributes to every float or double in all of our DBC files in the signal_valtype_list section. See [DBC File Format Documentation page 5-6](http://read.pudn.com/downloads766/ebook/3041455/DBC_File_Format_Documentation.pdf).

> Signals with value types 'float' and 'double' have additional entries in the signal_valtype_list section. 
> > signal_extended_value_type_list = 'SIG_VALTYPE_' message_id signal_name signal_extended_value_type ';' ;  

> >    signal_extended_value_type = '0' | '1' | '2' | '3' ; (* 0=signed or
unsigned integer, 1=32-bit IEEE-float, 2=64-bit IEEE-double *) 

I could also infer this from the scale factor and min/max values. Would prefer to do the former option as it is the correct way.

File: cantools/database/can/templates/DBC.h (hunk 3 above)
Comment (alisonbelow) [reply cantools/database/can/templates/DBC.h:243]: Yeah, all code in this class is autogenerated, buffer is read-only. Can add length but would not be used.

File: cantools/database/can/templates/DBC.h (hunk 4 above)
Comment (alisonbelow) [reply cantools/database/can/templates/DBC.h]: Fixed

**Pull request merged**